File: Game-Server/Parser/Lexer.py
from enum import Enum


class TokenType(Enum):
  EOF = 0
  String = 1
  LeftBracket = 2
  RightBracket = 3
  LeftParenthesis = 4
  RightParenthesis = 5
  Comma = 6

class Token:
  def __init__(self, type: TokenType, value: str = ""):
    self.type = type
    self.value = value

  def __str__(self):
    return f"{self.type}: {self.value}"

  def __repr__(self):
    return self.__str__()

  def __eq__(self, __o: object) -> bool:
    if isinstance(__o, Token):
      return self.type == __o.type and self.value == __o.value
    elif __o is TokenType:
      return self.type == __o
    elif self.type == TokenType.String and isinstance(__o, str):
      return self.value == __o
    return False

class Lexer:
  __ch: str

  def __init__(self, context):
    self.context = context
    self.pos = 0

  def scan(self) -> Token:
    self.read_char()
    self.ignore_space()
    if self.__ch == None:
      return Token(TokenType.EOF)
    elif self.__ch in ['"', "'"]:
      return Token(TokenType.String, self.get_string())
    elif self.__ch == '{':
      return Token(TokenType.LeftBracket)
    elif self.__ch == '}':
      return Token(TokenType.RightBracket)
    elif self.__ch == '(':
      return Token(TokenType.LeftParenthesis)
    elif self.__ch == ')':
      return Token(TokenType.RightParenthesis)
    elif self.__ch == ',':
      return Token(TokenType.Comma)
    ret = ""
    while self.__ch not in [' ' , '\t', '\n', '\r', '\f', \
                            '\v', None, ",", "}", ")", "{", "("]:
      if self.__ch == '\\':
        self.read_char()
        if self.__ch == '"':
          ret += '"'
        elif self.__ch == '\\':
          ret += '\\'
        elif self.__ch == 'n':
          ret += '\n'
        elif self.__ch == '(':
          ret += '('
        elif self.__ch == ')':
          ret += ')'
        elif self.__ch == 't':
          ret += '\t'
        elif self.__ch == '{':
          ret += '{'
        elif self.__ch == '}':
          ret += '}'
        elif self.__ch == 'r':
          ret += '\r'
        elif self.__ch == "'":
          ret += "'"
        else:
          ret += self.__ch
      else:
        ret += self.__ch
      self.read_char()
    if self.__ch != None:
      self.pos -= 1
    return Token(TokenType.String, ret)

  def get_string(self):
    ch = self.__ch
    ret = ""
    self.read_char()
    while self.__ch != None and self.__ch != ch:
      if self.__ch == '\\':
        self.read_char()
        if self.__ch == '"':
          ret += '"'
        elif self.__ch == '\\':
          ret += '\\'
        elif self.__ch == 'n':
          ret += '\n'
        elif self.__ch == 't':
          ret += '\t'
        elif self.__ch == 'r':
          ret += '\r'
        elif self.__ch == "'":
          ret += "'"
        else:
          ret += self.__ch
      else:
        ret += self.__ch
      self.read_char()
    return ret

  def ignore_space(self):
    while self.__ch != None and self.__ch in [' ', '\t', '\n', '\r', '\f', '\v']:
      self.read_char()

  def read_char(self, ch: str = None):
    if self.pos >= len(self.context):
      self.__ch = None
      return None
    self.__ch = self.context[self.pos]
    self.pos += 1

    # read_char(ch) -> bool
    if not isinstance(ch, str):
      return None
    if ch != self.__ch:
      return False
    self.__ch == ' '
    return True

File: src/lexer.cpp
#include "lexer.hpp"
#include "parser.hpp"
using namespace std;

Token Lexer::scan()
{
  read_char();
  ignore_space();
  if(m_ch == '\0')
    return Token(Token::TokenType::End);
  switch (m_ch)
  {
    case '\"':
    case '\'': return Token(get_string());
    case '(':  return Token(Token::TokenType::LeftParenthesis);
    case ')':  return Token(Token::TokenType::RightParenthesis);
    case ',':  return Token(Token::TokenType::Comma);
    case '{':  return Token(Token::TokenType::LeftBracket);
    case '}':  return Token(Token::TokenType::RightBracket);
    default:
      string ret;
      while(m_ch != ' '  && m_ch != '\t' && m_ch != '\n' 
        && m_ch != '\r' && m_ch != '\f' && m_ch != '\v'
        && m_ch != '{'  && m_ch != '}'  && m_ch != '('
        && m_ch != ')'  && m_ch != ','  && m_ch != '\0')
      {
        if(m_ch == '\\')
        {
          read_char();
          switch(m_ch)
          {
            default:
              ret += '\\';
              ret += m_ch;
              break;
            case '\0': ret += '\\';
              return Token(ret, Token::TokenType::String);
            case '\"': ret += '\"'; break;  
            case '\'': ret += '\''; break;
            case '\\': ret += '\\'; break;
            case 'n':  ret += '\n'; break;
            case 'r':  ret += '\r'; break;
            case 't':  ret += '\t'; break;
            case 'v':  ret += '\v'; break;
            case '{':  ret += '{';  break;
            case '}':  ret += '}';  break;
          }
        }
        else
          ret += m_ch;
        read_char();
      }
      // if(!m_ch)
      --m_pos;
      return Token(ret, Token::TokenType::String);
  }
}

string Lexer::get_string()
{
  string ret = "";
  char ch = m_ch;
  read_char();
  while(ch != '\0' && m_ch != ch)
  {
    if(m_ch == '\\')
    {
      read_char();
      switch(m_ch)
      {
        default:
          ret += '\\';
          ret += m_ch;
          break;
        case '\0': ret += '\\'; return ret;
        case '\"': ret += '\"'; break;
        case '\'': ret += '\''; break;
        case '\\': ret += '\\'; break;
        case 'n':  ret += '\n'; break;
        case 'r':  ret += '\r'; break;
        case 't':  ret += '\t'; break;
        case 'v':  ret += '\v'; break;
      }
    }
    else
      ret += m_ch;
    read_char();
  }
  return ret;
}

void Parser::block()
{
  // { stmt stmt* }
  match(Token::TokenType::LeftBracket);
  while(m_look != Token::TokenType::RightBracket)
    stmt();
  match(Token::TokenType::RightBracket);
}

void Parser::stmt()
{
  // block block*
  if(m_look == Token::TokenType::LeftBracket)
  {
    block();
    while(m_look == Token::TokenType::LeftBracket)
      block();
  }
  // command (, command)*
  else
  {
    command();
    while(m_look == Token::TokenType::Comma)
    {
      match(Token::TokenType::Comma);
      command();
    }
  }
}

void Parser::command()
{
  m_pcmd = new PCommand();
  id();
  // id(params)
  if(m_look == Token::TokenType::LeftParenthesis)
  {
    match(Token::TokenType::LeftParenthesis);
    params();
    match(Token::TokenType::RightParenthesis);
  }
  // id params
  else
    params();
  PCMDs.push_back(m_pcmd);
}

void Parser::params()
{
  if(m_look == Token::TokenType::String)
  {
    param();
    // param (, params)*
    if(m_look == Token::TokenType::Comma)
      while(m_look == Token::TokenType::Comma)
      {
        match(Token::TokenType::Comma);
        param();
      }
    // param params
    elif(m_look == Token::TokenType::String)
      while(m_look == Token::TokenType::String)
        param();
    // param
  }
  // ϵ
}

void Parser::param()
{
  auto tk = m_look;
  match(Token::TokenType::String);
  m_pcmd->AddParam(tk);
}

void Parser::id()
{
  auto id = m_look;
  match(Token::TokenType::String);
  char ch = id.value()[0];
  if(!isalpha(ch) && ch != '_')
    error("id must start with a letter or underscore");
  for(size_t i = 1; i < id.value().size(); ++i)
  {
    ch = id.value()[i];
    if(!isalnum(ch) && ch != '_')
      error("id must contain only letters, numbers, or underscores");
  }
  m_pcmd->SetFuncName(id.value());
}